### ISSUE TITLE
(Feature Request) Table Inheritance

### ISSUE THREAD

**Issue by tabennett** [analogueorm/analogue#101]
Hey, I was wondering if there is (currently) a way to implement table inheritance (single, join, etc) with Analogue, and if not, maybe we could hash out some ideas here. From looking at some of the underlying system classes, I would assume that this is not going to be trivial to implement. If there's anything I can do to help, please let me know. I think having a nice and clean way to do STI with Analogue would help a lot of people or are transitioning from AR to DM in their Laravel based apps and could really strengthen support for this project.

**Comment by RemiCollin** [comment]
I think it's already possible with a few tweaks using global scopes though I haven't tested it myself yet.

Analogue's mapping being class based, nothing prevents you from having a Admin Entity extending a User Entity, while sharing the same 'users' database table (you can extend the EntityMap accordingly). 

Then what you'd need is an additionnal column, let's say 'type', to use as a filter for queries. Then you'd apply a global scope to your Entity's mapper using this type column. 

Now, how it could looks like :

```php

class User extends Entity 
{

    public function __construct(...)
    {
         $this->type = 'user';
    }
}

class Admin extends User 
{
    public function __contruct(...)
    {
        $this->type = 'admin';
    }
}

class UserMap extends EntityMap 
{
    protected $table='users';

    public $type = 'user';
}

class AdminMap extends UserMap
{
    public $type = 'admin';
}

```

Now the global scope : 

```php

use Analogue\ORM\System\ScopeInterface;

class TypeScope implements ScopeInterface 
{
    public function apply($query)
    {
        $entityMap = $query->getMapper()->getEntityMap();
        
        $query->whereType($entityMap->type);
    }

    public function remove($query)
    {
        // Parse query and remove type filter
    }
}

```

Last, you need to register the scope on both User and Admin mappers, ideally somewhere in your application's service provider's boot method : 

```php

public function boot()
{
    $analogue = $this->app->make('analogue');
    $scope = new TypeScope;
    $analogue->mapper(User::class)->addGlobalScope($scope);
    $analogue->mapper(Admin::class)->addGlobalScope($scope);
}

```

**Comment by tabennett** [comment]
Man this should really be in the docs; I think it would help folks out tremendously. I'm going to test this out in a sample app I'm working on and see if it works. I'll post my results.

**Comment by RemiCollin** [comment]
Yes, please share your results. If this works, I'll code this as a plugin so it will make the workflow even easier. Reason it isn't in the doc is that it hasn't been tested yet.

**Comment by tabennett** [comment]
The code above works nicely for me. I have a simple example of STI created (using vehicles) and It's working nicely.

**Comment by tabennett** [comment]
Actually, this still won't provide true STI in the sense that (using your example above), you can't select all users (using the `UserMap`) and have it cast them into their respective child entity classes; It would cast them all into instances of `User` instead.

There doesn't seem to be a way to specify the entity class that you want to build/map during runtime (based upon that attributes being used to create the entity) inside of a map. 

I would like to propose that perhaps we could create a way to have the `Analogue\ORM\System\EntityBuilder::build` method check to see if the map has a discriminator column defined and if so, inject in value of that column into its call to `$this->mapper->newInstance()` (as part of the attributes array). Then, the base `Analogue\ORM\System\Mapper::newInstance` could pass that into its call to getClass like this:

```php
$this->entityMap->getClass($attributes);
```

Then, a user could just override the `getClass` method on their entity map and use the discriminator value to determine how what class we want to map the entity to. Thoughts?

**Comment by RemiCollin** [comment]
Currently there is a one-one equallity on mappers and entity classes so it may raises several issues. I'll see how we can implement this

**Comment by RemiCollin** [comment]
Actually, it might be easier that i first thought. I'll try to implement this in the upcoming 5.3 release.

**Comment by tabennett** [comment]
I would LOVE to be able to implement table inheritance using Analogue; right now I'm building a pretty complex billing application using DDD with Analogue. Up to this point it's been working pretty well for me, but I've got to be able to have inheritance (class, single table) in my DB tables in order to have this working properly and not have to use Doctrine (ugh). If you want some help implementing this I'm more than willing. Let me know if there's anything I can do to help.

**Comment by RemiCollin** [comment]
Sure, if you want to have a take on this, please do. You can create a feature branch from the `5.3-dev` which is the current dev branch.

I have a few ideas on how it could work :

First I think we'll need some refactoring and to add an upper layer to the EntityBuilder class, something like a `ResultBuilder` class, that would instantiate as many EntityBuilder as needed to build a result set. The EntityBuilder class itself should then be modified to take only a single record as an input. 

On our base EntityMap we'd use a `$discriminatorColumn` property which would tell the `ResultBuilder` which class the Entity belongs. In our previous example, this discriminator would be set on the `UserMap`. In that case this `type` column could contains the full namespace of the entityClass (in a similar mechanism as polymorphic relationships), which would make it fairly easy to deal with. What do you think ?

Let me know if you have any question.

**Comment by tabennett** [comment]
Hey I've been working on this, but I'm not quite sure about a couple of things. In the `ResultBuilder` class, we'll need to be able to keep track of an array of `EntityBuilder` instances as we spin through the results and create entities from each one. However, in order to do that we'll need to have mappers for each `EntityBuilder` that we create. What are your thoughts on a clean way to do this?

**Comment by RemiCollin** [comment]
I think we'd just create `EntityBuilder` classes on-demand and store them in an associative array, keyed by entity class, in the `ResultBuilder`.
So as we spin our result set we'd check if the `EntityBuilder` for the current Entity::class exists in our array, and if it's not instantiated yet, we'd call a `new EntityBuilder(Manager::mapper(Entity::class))`.

**Comment by RemiCollin** [comment]
Implemented in 5.3
